src/ParticleContainer.cpp

```diff
@@ -54,12 +42,365 @@ std::vector<Particle>::const_iterator ParticleContainer::end() const {
   return particles.end();
 }
 
-void ParticleContainer::addCube(std::vector<Particle> particleCube) {
+void ParticleContainer::handleLJFCalculation() {}
+
+void ParticleContainer::resetParticles() { particles.clear(); }
+
+void ParticleContainer::addMultipleParticles(std::vector<Particle> particleCube) {
   for (size_t x = 0; x < particleCube.size(); ++x) {
     addParticle(particleCube.at(x));
   }
 }
 
+void LCParticleContainer::setBoundarys(std::array<int, 6> in) {
+  boundary_types = in;
+}
+
 void ParticleContainer::setParticle(Particle p, int position) {
   particles.at(position) = p;
 }
+
+std::vector<Particle>
+LCParticleContainer::getParticleInNeighbourhood(std::array<int, 3> id) {
+  std::vector<Particle> neigbourhood;
+  int particleCount = 0;
+  for (auto &cell : cells) {
+    if (cell.getParticles().size() > 0) {
+      int x = cell.getId().at(0) - id.at(0);
+      if (x <= 1 && x >= -1) {
+        int y = cell.getId().at(1) - id.at(1);
+        if (y <= 1 && y >= -1) {
+          int z = cell.getId().at(2) - id.at(2);
+          if (z <= 1 && z >= -1) {
+            if (x != 0 || y != 0 || z != 0) {
+              for (auto p : cell.getParticles()) {
+                neigbourhood.push_back(*p);
+                particleCount++;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  // spdlog::info("neighbourhood has {} particles", particleCount);
+  if (particleCount <= 0) {
+    return {};
+  } else {
+    //spdlog::info("there was a neigbourhood of {} particles", particleCount);
+    return neigbourhood;
+  }
+}
+
+Cell &LCParticleContainer::getCellById(std::array<int, 3> id) {
+  if (cells.size() != 0) {
+    for (auto &c : cells) {
+      if (c.getId() == id) {
+        return c;
+      }
+    }
+  }
+}
+
+void LCParticleContainer::realocateParticles() {
+  //spdlog::info("we have {} cells and {} particles", cells.size(),getParticles().size());
+  for (auto &c : cells) {
+    c.emptyCell();
+  }
+  std::vector<Particle> updatedParticles;
+  for (auto &p : particles) {
+    int x = floor(p.getX().at(0) / cell_size.at(0));
+    int y = floor(p.getX().at(1) / cell_size.at(1));
+    int z = floor(std::abs(p.getX().at(2)) / cell_size.at(2));
+    p.setF({0.0, 0.0, 0.0});
+    if (cellExists({x, y, z})) {
+      getCellById({x, y, z}).addParticle(&p);
+      updatedParticles.push_back(p);
```

Comment:
Every timestep, you copy all particles. This is a huge amount of memory consumption.
Actually, this might also be your source of a main bug. Because your particle pointers are no longer valid after copying the particle.

Reply:
A simple solution would be to split outflow handling and restructuring:
```cpp
for (auto &p : particles) {
  if (p outside domain) {
    remove the particle
}}
for (auto &p : particles) {
   getCellById().addParticle(&p);
}
```

Reply:
Dont quite get the point here. At the Moment we are not actually deleting the Particles, we are simply not adding them to the updatedParticles. In the first loop over all cells we are clearing all Pointers. In case you mean this with the invalidation of our Pointers, I dont know a way around it.

Reply:
The detail that is easy to miss here, is that `push_back()` creates a copy of the given object.
So, what happens in the inner `if` clause is this:
` getCellById({x, y, z}).addParticle(&p);` -> You add a pointer to the particle p to the corresponding cell.
` updatedParticles.push_back(p);`  -> You add a copy of `p` to `updatedParticles`.

In the end, you set `particles` to be the `updatedParticles`. So, your `particles` vector contains all the copies from the particles, while the particle pointers inside the cells still point to the original particles.

Reply:
`if (cellExists({x, y, z})) {
      updatedParticles.push_back(p);
      getCellById({x,y,z})->addParticle(&updatedParticles.back());
    }`Now i get your Point and i didnt had a clue about Push_back storing only copies. After a quick online search for alternatives for Push_back without any results, i changed the order of the two lines. Now we store Pointers to the newly created Particle Copies in our Cells. I think that eliminates the double Particle problem